spam/spam_tutorial.py

```diff
@@ -140,234 +153,407 @@
+# not to label everything or directly create a model for inference using the LFs.
+# The training labels are used to train a separate discriminative model (in this case, one which just uses the comment text) in order to generalize to new, unseen examples.
+# Using this model, we can make predictions for data points that our LFs don't cover.
+#
+# We'll walk through the development of two LFs using basic analysis tools in Snorkel, then provide a full set of LFs that we developed for this tutorial.
 
 # %% [markdown]
-# ### a) Look at examples for ideas
+# ### a) Initial ideation: look at training examples
 
 # %% [markdown]
-# We begin the process of writing LFs by looking at some examples in the dev set.
+# Following recommended practice above, we'll start by looking at the `train` set to generate some ideas for LFs.
 
 # %%
-# Display just the text and label
-df_dev[["text", "label"]].sample(10, random_state=3)
+df_train[["author", "text", "video"]].sample(20, random_state=2)
+
+# %% [markdown]
+# One dominant pattern in the comments that look like spam is the use of the phrase "check out" (e.g. "check out my channel").
+# Let's start with that.
 
 # %% [markdown]
-# ### b) Writing an LF
+# ### b) Writing an LF to identify spammy comments that use the phrase "check out"
 
 # %% [markdown]
 # Labeling functions in Snorkel are created with the `@labeling_function()` decorator, which wraps a function for evaluating on a single data point (in this case, a row of the DataFrame).
+# Decorators are a convenient Python mechanism for automatically extending the behavior of the decorated function, and you can learn more [here](https://realpython.com/primer-on-python-decorators/).
 #
-# Looking at samples of our data, we see multiple messages where spammers are trying to get viewers to look at "my channel" or "my video," so we write a simple LF that labels an example as `SPAM` if it includes the word "my" and otherwise abstains.
+# Let's start developing an LF to catch instances of commenters trying to get people to "check out" their channel, video, or website.
+# We'll start by just looking for the exact string `"check out"` in the text, and compare that to looking for just `"check"` in the text.
+# We'll write a Python function over a single data point (here, a [Pandas `Series` object](https://pandas.pydata.org/pandas-docs/stable/reference/series.html)) to express the two versions of our rule, then add the decorator.
 
 # %%
 from snorkel.labeling.lf import labeling_function
 
 
 @labeling_function()
-def keyword_my(x):
-    """Many spam comments talk about 'my channel', 'my video', etc."""
-    return SPAM if "my" in x.text.lower() else ABSTAIN
+def check(x):
+    return SPAM if "check" in x.text.lower() else ABSTAIN
 
 
-lfs = [keyword_my]
+check
+
+
+# %%
+@labeling_function()
+def check_out(x):
+    return SPAM if "check out" in x.text.lower() else ABSTAIN
+
+
+check_out
```

Comment:
I don't think adding anything to print out the LF like this here or below, would cut